Add capability to run forecast in segments

WalterKolczynski-NOAA opened the pull request:
# Description
Adds the ability to run a forecast in segments instead of all at once. To accomplish this, a new local `checkpnts` variable is introduced to `config.base` to contain a comma-separated list of intermediate stopping points for the forecast. This is combined with `FHMIN_GFS` and `FHMAX_GFS` to create a comma-separated string `FCST_SEGMENTS` with all the start/end points that is used by `config.fcst` and rocoto workflow. Capability to parse these into python lists was added to wxflow in an accompanying PR. If `checkpnts` is an empty string, this will result in a single-segment forecast.

To accommodate the new segment metatasks that must be run serially, the capability of `create_task()` was expanded to allow a dictionary key of `is_serial`, which controls whether a metatask is parallel or serial using pre-existing capability in rocoto. The default when not given is parallel (i.e. most metatasks).

Resolves #2274
Refs NOAA-EMC/wxflow#39
Refs NOAA-EMC/wxflow#40

# Type of change
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
- Forecast-only on Hercules
- GEFS on Hercules

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

WalterKolczynski-NOAA committed:
Add capability to run forecast in segments

Adds the ability to run a forecast in segments instead of all at
once. To accomplish this, the `FHMIN_GFS` and `FHMAX_GFS` settings
have been replaced as user-setable variables in favor of
`FCST_SEGMENTS_STR_GFS`, a comma-separated list of the segment
boundaries (thus there will be one more than the number of segments).
For a traditional single-segment forecast, this would just be set to
`"${FHMIN_GFS},${FHMAX_GFS}"`.

The comma-separated list had to be used instead of a bash array as
the variable must be exported in order for the rocoto generator to
see it, and arrays cannot be exported from shell. Capabilty to parse
these into python lists was added to wxflow in an accompanying PR.

To accomodate the new segment metatasks that must be run serially,
the capability of `create_task()` was expanded to allow a dictionary
key of `is_serial`, which controls whether a metatask is parallel or
serial using pre-existing capability in rocoto. The default when not
given is parallel (i.e. most metatasks).

Resolves #2274
Refs NOAA-EMC/wxflow#39

WalterKolczynski-NOAA commented:
I still have some more testing to do (and I think documentation to update), but wanted to get a draft out since people are going on leave.

WalterKolczynski-NOAA committed:
Fix pycode style

WalterKolczynski-NOAA committed:
Clean up remaining bugs in fcst segment update

WalterKolczynski-NOAA commented:
Will check and see if the documentation needs to be updated Friday, but I think the last few bugs from this update are gone.

DavidHuber-NOAA commented on parm/config/gefs/yaml/defaults.yaml:
```diff
@@ -9,7 +9,7 @@ base:
   KEEPDATA: "NO"
   DO_EXTRACTVARS: "NO"
-  FHMAX_GFS: 120
+  FCST_SEGMENTS_GFS: "0,48,120"
```
Just for my own understanding, does this mean run a forecast from f000 through f048, then another from f048 through f120?

WalterKolczynski-NOAA replied:
Yes.

DavidHuber-NOAA commented:
Hercules testing completed successfully, resetting label.

WalterKolczynski-NOAA commented:
Appears no documentation updates are needed at this time after all.

aerorahul commented on parm/config/gfs/config.base:
```diff
@@ -289,8 +289,19 @@ export EUPD_CYC="@EUPD_CYC@"
-export FHMAX_GFS=@FHMAX_GFS@
+# Comma-separated forecast hour intervals to run the forecast over
+#   For a single-segment forecast, this is simply "$FHMIN_GFS,$FHMAX_GFS"
+export FCST_SEGMENTS_STR_GFS="@FCST_SEGMENTS_GFS@"
+IFS=', ' read -ra FCST_SEGMENTS_GFS <<< "${FCST_SEGMENTS_STR_GFS}"
+if (( ${FCST_SEGMENT:- -1} < 0 )); then
+    # Jobs other than the forecast don't care about segments, only the
+    #   absolute start and end
+    declare -x FHMIN_GFS=${FCST_SEGMENTS_GFS[0]}
+    declare -x FHMAX_GFS=${FCST_SEGMENTS_GFS[-1]}
+else
+    declare -x FHMIN_GFS=${FCST_SEGMENTS_GFS[${FCST_SEGMENT}]}
+    declare -x FHMAX_GFS=${FCST_SEGMENTS_GFS[${FCST_SEGMENT}+1]}
+fi
```
Is there a way to not do any of this in a config file and calculate this in a j-job or exscript?

WalterKolczynski-NOAA replied:
Not without massive additional changes. `FHMAX_GFS` especially gets used later in this config, and then also in the job-specific configs that would be sourced immediately afterwards.

WalterKolczynski-NOAA replied:
Re-done how we discussed Mon afternoon.

WalterKolczynski-NOAA committed:
Change fcst segments to breakpoints, move to config.fcst

Changes the way forecast segments are defined. Restores the original
`FHMIN_GFS` and `FHMAX_GFS` and then adds a local `breakpnts` variable
that contains the intermediate stopping points (if any). The original
list of segment endpoints is then constructed from that.

The determination of the `FHMIN` and `FHMAX` based on the segment is
moved from `config.base` to `config.fcst`. This required adding some
additional checks in `config.fcst` to clip other `FHMAX` variables to
`FHMAX`.

KateFriedman-NOAA reviewed (dismissed):
Looks good, thanks @WalterKolczynski-NOAA !

EricSinsky-NOAA reviewed (approved):
Thank you for adding this capability, Walter. All these changes look good to me.

DavidHuber-NOAA commented:
Can you update your branch @WalterKolczynski-NOAA?

WalterKolczynski-NOAA commented:
> Can you update your branch @WalterKolczynski-NOAA?

Thought I had, but had another conflict in the rebase so it didn't push. Should be right now.

aerorahul reviewed (dismissed):
looks good.
I have not tested or inspected the XML that comes out of the setup, but will try and do that sometime soon.

WalterKolczynski-NOAA committed:
Update Test for GFS forecast segments

Updates the extended GFS case to use forecast segments to test that
capability for the GFS system (the GEFS case already tests segments
as well).

WalterKolczynski-NOAA commented:
Failures are due to the change that automatically parses comma separated values in configs as python lists. I've made a very small follow-on PR to wxflow to remove empty elements, which will allow us to create single-element lists.

WalterKolczynski-NOAA committed:
Adjust aerosol DA jobs for new config arrays

Earlier update to parse comma-separated bash variables as lists in
python mean we no longer need to do that in the task scripts.

This commit is incomplete until a follow-up PR that will update the
wxflow hash.

WalterKolczynski-NOAA committed:
Update wxflow to complete changes made in previous commit

WalterKolczynski-NOAA committed:
Update archive for fcst segments

WalterKolczynski-NOAA committed:
Update archive jinja to account for new lists

With comma-separated lists now being read into python as lists, the
jinja templates for the archive job had to be updated to not attempt
to create them anymore.

DavidHuber-NOAA commented on parm/archive/gfs_netcdfa.yaml.j2:
```diff
@@ -6,7 +6,7 @@ gfs_netcdfa:
         - "{{ COMIN_ATMOS_ANALYSIS | relpath(ROTDIR) }}/{{ head }}sfcanl.nc"
         - "{{ COMIN_ATMOS_ANALYSIS | relpath(ROTDIR) }}/{{ head }}atminc.nc"
-        {% for iauhr in iaufhrs if iauhr != 6 %}
+        {% for iauhr in IAUFHRS if iauhr != 6 %}
```
For 3DVar, `IAUFHRS=6`, so it will not be a list when parsed by wxflow.  Instead, you may need an additional check here:

```suggestion
        {% if IAUFHRS is not integer %} # If it is, then there will only be one increment file (handled above)
            {% for iauhr in IAUFHRS if iauhr != 6 %}
```

DavidHuber-NOAA commented on parm/archive/gdas_restarta.yaml.j2:
```diff
@@ -6,7 +6,7 @@ gdas_restarta:
         - "{{ COMIN_ATMOS_ANALYSIS | relpath(ROTDIR) }}/{{ head }}atminc.nc"
         # IAU increments
-        {% for iaufhr in iaufhrs if iaufhr != 6 %}
+        {% for iaufhr in IAUFHRS if iaufhr != 6 %}
```
```suggestion
        {% if IAUFHRS is not integer %} # If it is, then there will only be one increment file (handled above)
            {% for iauhr in IAUFHRS if iauhr != 6 %}
```

DavidHuber-NOAA replied:
I think I prefer the master option.  Resolving.

DavidHuber-NOAA commented on parm/archive/master_gfs.yaml.j2:
```diff
@@ -3,18 +3,6 @@
 {% set cycle_YMD = current_cycle | to_YMD %}
 {% set cycle_YMDH = current_cycle | to_YMDH %}
 
```
Alternatively, you could just add the check to the masters and convert `int`s to `list`s.

```suggestion

{% if IAUFHRS is integer %}
    {% set IAUFHRS = [IAUFHRS] %}
{% endif %}

```

DavidHuber-NOAA commented on parm/archive/master_gdas.yaml.j2:
```diff
@@ -3,16 +3,6 @@
 {% set cycle_YMDH = current_cycle | to_YMDH %}
 {% set head = "gdas.t" + cycle_HH + "z." %}
 
```
```suggestion

{% if IAUFHRS is integer %}
    {% set IAUFHRS = [IAUFHRS] %}
{% endif %}

```

DavidHuber-NOAA commented on parm/archive/master_enkf.yaml.j2:
```diff
@@ -4,28 +4,6 @@
 {% set cycle_YMDH = current_cycle | to_YMDH %}
 {% set head = RUN + ".t" + cycle_HH + "z." %}
 
```
```suggestion

{% if IAUFHRS is integer %}
    {% set IAUFHRS = [IAUFHRS] %}
{% endif %}

{% if IAUFHRS_ENKF is integer %}
    {% set IAUFHRS_ENKF = [IAUFHRS_ENKF] %}
{% endif %}

```

WalterKolczynski-NOAA replied:
I think this won't be necessary once I do the thing I intended to do, which is add a trailing comma to the single-digit IAUFHRS.

WalterKolczynski-NOAA committed:
Add trailing commas when IAUFHRS is only one number

WalterKolczynski-NOAA committed:
Fix cdump to run and make AUFHRS always array

WalterKolczynski-NOAA committed:
Fix wxflow import

WalterKolczynski-NOAA committed:
Correct prod dependency

Missed updating one of the dependencies from task to metatask.

WalterKolczynski-NOAA committed:
Fix another dependency bug

WalterKolczynski-NOAA commented:
Experiments are actually still running

TerrenceMcGuinness-NOAA commented:
All Cases **Passed** on Hercules

```
mterry (hercules-login-4) RUNTESTS $ pwd
/work2/noaa/stmp/CI/HERCULES/2795/RUNTESTS

mterry (hercules-login-4) RUNTESTS $ cat ci-run_check.log 
Experiment C48_ATM_d0927f02 Completed 1 Cycles: *SUCCESS* at Wed Aug  7 22:55:06 CDT 2024
Experiment C96C48_hybatmDA_d0927f02 Completed 3 Cycles: *SUCCESS* at Wed Aug  7 23:55:47 CDT 2024
Experiment C96_atm3DVar_d0927f02 Completed 3 Cycles: *SUCCESS* at Thu Aug  8 00:01:41 CDT 2024
Experiment C48_S2SW_d0927f02 Completed 1 Cycles: *SUCCESS* at Thu Aug  8 00:56:06 CDT 2024
Experiment C48_S2SWA_gefs_d0927f02 Completed 1 Cycles: *SUCCESS* at Thu Aug  8 01:15:09 CDT 2024
```

Manually setting state label for Hercules to PASSED

WalterKolczynski-NOAA commented:
The extended test is working when I try manually, so I am going to attempt CI again.

WalterKolczynski-NOAA committed:
Only delete temp restarts if it is the final fcst segment

WalterKolczynski-NOAA committed:
Fix default syntax

DavidHuber-NOAA commented on parm/config/gfs/config.aeroanl:
```diff
@@ -24,7 +24,7 @@ if [[ "${DOIAU}" == "YES" ]]; then
   export JEDIYAML="${PARMgfs}/gdas/aero/variational/3dvar_fgat_gfs_aero.yaml.j2"
 else
-  export aero_bkg_times="6"
+  export aero_bkg_times="6," # Trailing comma is necessary so this is treated as a list
```
Very clever!

DavidHuber-NOAA reviewed (approved):
Looks good to me.
